src/api/app/controllers/trigger_controller.rb

```diff
@@ -1,53 +1,96 @@
 class TriggerController < ApplicationController
+  validate_action rebuild: { method: :post, response: :status }
+  validate_action release: { method: :post, response: :status }
   validate_action runservice: { method: :post, response: :status }
 
+  before_action :extract_auth_from_request, :validate_auth_token, :require_valid_token
+
+  before_action :require_project_param, only: [:release]
   #
-  # This controller is checking permission always only on the base of tokens
+  # Authentication happens with tokens, so no login is required
   #
   skip_before_action :extract_user
   skip_before_action :require_login
 
-  # github.com sends a hash payload
-  skip_before_action :validate_params, only: [:runservice]
+  def rebuild
+    Backend::Api::Sources::Package.rebuild(@pkg.project.name, @pkg.name)
+    render_ok
+  end
 
-  def runservice
-    auth = request.env['HTTP_AUTHORIZATION']
+  # rubocop:disable Lint/NonLocalExitFromIterator
+  def release
+    matched_repo = false
+    User.session = @token.user
+    @pkg.project.repositories.includes(:release_targets).each do |repo|
+      repo.release_targets.where(trigger: 'manual').each do |releasetarget|
+        unless @token.user.can_modify?(releasetarget.repository.project)
```

Comment:
If I am not mistaken, we want to check the `target_repository` of the release target instead: `releasetarget.target_repository.project`

Reply:
right, this is indeed wrong

Reply:
@adrian 

so i have to check the permission on `releasetarget.target_repository.project` ?

what else is wrong? the call to `release_package` is right or there is something wrong there too?